src/DS/sds.c

```diff
@@ -246,108 +270,179 @@ static int ds_sds_dump_component(const char* external_file, const char* componen
 		return -1;
 	}
 
-
-	const char* component_id;
-	const char* filename;
-
 	if (xlink_href[0] == '#')
 	{
-		filename = NULL;
-		component_id = xlink_href + 1;
+		*component_id = xlink_href + 1;
+
+		return ds_sds_dump_local_component(*component_id, session, target_filename_dirname, relative_filepath);
 
 	} else if (oscap_str_startswith(xlink_href, "file:")){
 
 		char* sep = strchr(xlink_href, '#');
 
-		filename = xlink_href + strlen("file:");
+		const char *filename = xlink_href + strlen("file:");
 
 		if (sep == NULL) {
-			component_id = NULL;
+			*component_id = NULL;
 		} else {
 			*sep = '\0';
-			component_id = sep + 1;
+			*component_id = sep + 1;
 		}
 
+		return ds_sds_dump_file_component(filename, *component_id, session, target_filename_dirname, relative_filepath);
+
+	} else if (oscap_acquire_url_is_supported(xlink_href)){
+		char *sep = strchr(xlink_href, '#');
+
+		char* url = xlink_href;
 
+		if (sep == NULL) {
+			*component_id = NULL;
+		} else {
+			*sep = '\0';
+			*component_id = sep + 1;
+		}
+
+		if (!ds_sds_session_fetch_remote_resources(session)) {
+			static bool fetch_remote_resources_suggested = false;
+
+			if (!fetch_remote_resources_suggested) {
+				fetch_remote_resources_suggested = true;
+				ds_sds_session_remote_resources_progress(session)(false, "Datastream component '%s' points out to the remote '%s'. "
+									"Use '--fetch-remote-resources' option to download it.\n", cref_id, url);
+			}
+
+			ds_sds_session_remote_resources_progress(session)(true, "WARNING: Skipping '%s' file which is referenced from datastream\n", url);
+			return -2;
+		}
+
+		return ds_dsd_dump_remote_component(url, *component_id, session, target_filename_dirname, relative_filepath);
 	} else {
 		oscap_seterr(OSCAP_EFAMILY_XML, "Unsupported type of xlink:href attribute on given component-ref - '%s'.", xlink_href);
-		xmlFree(cref_id);
-		xmlFree(xlink_href);
 		return 0;
 	}
+	return 0;
+}
 
-	char* filename_cpy = oscap_sprintf("./%s", relative_filepath);
-	char* file_reldir = dirname(filename_cpy);
+int ds_sds_dump_component_ref_as(const xmlNodePtr component_ref, struct ds_sds_session *session, const char* sub_dir, const char* relative_filepath)
+{
+	char* cref_id = (char*)xmlGetProp(component_ref, BAD_CAST "id");
+	if (!cref_id)
+	{
+		oscap_seterr(OSCAP_EFAMILY_XML, "No or invalid id attribute on given component-ref.");
+		xmlFree(cref_id);
+		return -1;
+	}
 
-	const char* target_filename_dirname = oscap_sprintf("%s/%s",sub_dir, file_reldir);
-	oscap_free(filename_cpy);
+	char* xlink_href = (char*)xmlGetNsProp(component_ref, BAD_CAST "href", BAD_CAST xlink_ns_uri);
 
-	ds_sds_dump_component(filename, component_id, session, target_filename_dirname, relative_filepath);
+	char* target_filename_dirname = compose_target_filename_dirname(relative_filepath, sub_dir);
 
+	char* component_id = NULL;
+
+	int ret = ds_sds_dump_component_by_href(session, xlink_href, target_filename_dirname, relative_filepath, cref_id, &component_id);
```

Comment:
This isn't much important, but I have noticed one thing while doing the review.

I think that `cref_id` and `xlink_href` are not used after this line anymore.
What if you put here this
```
xmlFree(cref_id);
xmlFree(xlink_href);
```
and remove those from the rest of the function? It's 6 times there.